Change Statistics model and basic endpoints

pbylicki opened the pull request.

MaciejWiczk committed:
Fix SQLAlchemy dependency in setup.py

MaciejWiczk committed:
Adding early changes for endpoints and models

MaciejWiczk committed:
Merge branch 'master' of https://github.com/pbylicki/rfhub2 into feature/add-statics-of-keyword-usage

MaciejWiczk committed:
adding missing file

pbylicki committed:
Change Statistics model and basic endpoints

pbylicki committed:
Add aggregated statistics endpoints

pbylicki committed:
Refactor statistics filter params

pbylicki committed:
Add statistics time range tests

pbylicki committed:
Fix avg_elapsed calculation

MaciejWiczk committed:
Add statistics to cli (#124)

* Extended cli, api_client, rfhub_importer with statics

* Fixing comments and aggregated query

* Blacked

* Refactor and some tests added, still need one to pass

* Adding missing files

* More tests

* adding ignored files

* More tests added

* Fixing tests

* Fixing tests

* Fixing tests

* Fixing tests

* Extendind tests and changind README.md to align with code changes

* Final touches before PR review

* PR comments applied

* Forgotten dict

MaciejWiczk commented:
@pbylicki what do we do about this one? Should we merge it to master to avoid conflicts later on, or will You work on it, and merge should wait?

pbylicki commented:
@MaciejWiczk I wouldn't merge it yet since there still may be some changes in the API and we will be releasing some smaller changes in the meantime. In particular I would like to create and test the endpoints that would return collections and keywords with statistics included (aggregated per request, not cached yet) to see if it can be used as initial implementation and base for UI changes.

pbylicki committed:
Merge remote-tracking branch 'origin/master' into change_statistics_model_and_endpoints

pbylicki committed:
Update deprecated method calls

pbylicki committed:
Add collection and keyword get endpoints that calculate statistics

MaciejWiczk committed:
Merge pull request #137 from pbylicki/add-stats-to-collection-keyword-get-results

Add collection and keyword get endpoints that calculate statistics

MaciejWiczk committed:
Adding times_used to CollectionList (#138)

pbylicki committed:
Show keyword statistics on CollectionDetails view (#139)

* Add filtering keywords by collection_id

* Show keyword statistics on CollectionDetails view

MaciejWiczk commented on rfhub2/api/endpoints/keywords.py:
```diff
@@ -32,9 +39,14 @@ def get_keywords_with_stats(
 ):
     return repository.get_all_with_stats(
-        skip=skip, limit=limit, pattern=pattern, use_doc=use_doc
+        skip=skip,
```
Sorry for not noticing it earlier. Due to different endpoint app is showing only first 100 keywords. Maybe when collection_id is not None we could skip the limit, to have same experience as previosly?

pbylicki replied:
good point, I will set UI to use large limit for now

MaciejWiczk committed:
Adding Limiti=10000 to client request to show all available keywords (#150)

pbylicki committed:
Add sorting support to all sortable endpoints (#151)

* Add ordering to existing API resources

* Add ordering tests

MaciejWiczk commented:
@pbylicki what is preventing from merging this one? Do You have some other plans for that one?

pbylicki commented:
@MaciejWiczk did you have a chance to test it with some real life data to see how adding the statistics calculation affects the performance?
Apart from merging all the latest changes I would rename Statistics to KeywordStatistics across the app to be more specific and prepare for different kinds of statistics in the future

MaciejWiczk commented:
> @MaciejWiczk did you have a chance to test it with some real life data to see how adding the statistics calculation affects the performance?

I'm on holidays till next friday, so it will have to wait till then. 

> Apart from merging all the latest changes I would rename Statistics to KeywordStatistics across the app to be more specific and prepare for different kinds of statistics in the future.

Sounds very reasonable. I could create PR with those changes and I will resolve conflicts on the way.

MaciejWiczk committed:
Change statistics model and endpoints merge (#165)

* Bump fastapi from 0.45.0 to 0.46.0

Bumps [fastapi](https://github.com/tiangolo/fastapi) from 0.45.0 to 0.46.0.
- [Release notes](https://github.com/tiangolo/fastapi/releases)
- [Changelog](https://github.com/tiangolo/fastapi/blob/master/docs/history-design-future.md)
- [Commits](https://github.com/tiangolo/fastapi/compare/0.45.0...0.46.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

* Bump coverage from 5.0.2 to 5.0.3

Bumps [coverage](https://github.com/nedbat/coveragepy) from 5.0.2 to 5.0.3.
- [Release notes](https://github.com/nedbat/coveragepy/releases)
- [Changelog](https://github.com/nedbat/coveragepy/blob/master/CHANGES.rst)
- [Commits](https://github.com/nedbat/coveragepy/compare/coverage-5.0.2...coverage-5.0.3)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

* Fixes for #152, #153, #155. (#156)

* Fixes for #152, #153, #155. * Extending Docs by init method docs, when init parametrised * * Allow importing robot filesform directory with library with __init__ * Printing loaded collections alphabetically

* Review changes

* Bump fastapi from 0.46.0 to 0.47.1 (#157)

Bumps [fastapi](https://github.com/tiangolo/fastapi) from 0.46.0 to 0.47.1.
- [Release notes](https://github.com/tiangolo/fastapi/releases)
- [Changelog](https://github.com/tiangolo/fastapi/blob/master/docs/history-design-future.md)
- [Commits](https://github.com/tiangolo/fastapi/compare/0.46.0...0.47.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

* Added elipsis to long keyword names and tooltip to keywords(#133) (#144)

* Added elipsis to long keyword names and tooltip to keywords(#133)

* Changing locator

* Fixes based on comments(#133).

* Fix for failing test

Co-authored-by: Maciej Wiczk <35737166+MaciejWiczk@users.noreply.github.com>

* Remove search character limit (#164)

Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>
Co-authored-by: MarcinMaciaszek <50177134+MarcinMaciaszek@users.noreply.github.com>
Co-authored-by: Paweł Bylicki <bylak@poczta.fm>

pbylicki commented:
@MaciejWiczk Renaming resolved in #166

pbylicki committed:
Rename Statistics resource to KeywordStatistics (#166)

* Rename Statistics resource to KeywordStatistics

* Update robot tests

MaciejWiczk committed:
Resolve conflicts

pbylicki committed:
Merge pull request #167 from pbylicki/change_statistics_model_and_endpoints_2

Change statistics model and endpoints conflicts resolution

pbylicki committed:
Statistics extractor refactoring (#168)

* Statistics extractor refactoring

* Unify post request body serialization

* Update fail scenario

MaciejWiczk commented:
@pbylicki I did some tests on "prod" env in company and I've noticed one thing: Posts requests are getting slower and slower with every row there is added into table in DB. I believe it is happening due to recalculation of indices. One option for mitigation would be to add an endpoint, that would accept more than one statistic object, or we could block indices recalculation on that table. I think that first approach would be nicer, but I will investigate indices recalculation as well.

pbylicki commented:
@MaciejWiczk Thanks for testing, did you see any issues with response time of endpoints reading collections and keywords with aggregated statistics?
I initially wanted to create endpoint for adding many statistics record, but it was lost at some point. Thanks for reminding me. I will add this before merging this PR. It should partially address the slowness. You are right that index on this table is to blame.

MaciejWiczk commented:
> did you see any issues with response time of endpoints reading collections and keywords with aggregated statistics?

frankly speaking I didn't think about it, but I will give it a go when first occasion occurs. 

> You are right that index on this table is to blame.

Manually adding all the record by SQL IDE, either in one statement or multiple single row inserts didn't cause this behaviour. But this is probably by IDE design tough.

MaciejWiczk commented:
> > did you see any issues with response time of endpoints reading collections and keywords with aggregated statistics?
> 
> frankly speaking I didn't think about it, but I will give it a go when first occasion occurs.
> 
> > You are right that index on this table is to blame.
> 
> Manually adding all the record by SQL IDE, either in one statement or multiple single row inserts didn't cause this behaviour. But this is probably by IDE design tough.

So I did tests on corporate network with fair amount of keywords. This time, having app running on localhost and Postgres DB running on OC turned out to behave linearly. 1.5 GB of outputs.xml were transformed into  ~37k rows in DB in about 3 hours. Maybe previous time I spot some net bottleneck that was fixed. To wrap up, this time it's behaving correctly, data like libraries, keywords are loading the same way they did, with the same speed. Statistics could be loaded via some endpoint that allows batch loading, as You mentioned before. I think that adding some progressbar would be beneficial for user, to indicate that app is working correctly and adding data. I will create PR with that.

pbylicki commented:
@MaciejWiczk I have a branch with endpoint for adding batch statistics, just need to update cli part to use it. Should be ready this week

MaciejWiczk commented:
> Should be ready this week.

Awesome, I can take care of cli. Can You check my PR #180 for merging it into yours? It will be simpler to work on synchronized branch.

MaciejWiczk committed:
Change statistics model and endpoints typing (#180)

* add typing to keywords_importer.py

* Adding dataclass for most common CollectionUpdate, List[KeywordUpdate] combination

* Fix cli tests (#191)

* Fix robot CLI options tests

* Blacked file

* Fixed typings

* PR fixes

Co-authored-by: Paweł Bylicki <pawelkbylicki@gmail.com>

pbylicki committed:
Add many statistics endpoint (#195)

* Add many keyword statistics endpoint

* Adjust CLI to new endpoint

* Formatting

MaciejWiczk commented:
@pbylicki, could You bump version before merging that one? Or do we have to do something before that?

pbylicki commented:
@MaciejWiczk I want to make client upload statistics in batches before the merge

MaciejWiczk commented:
> @MaciejWiczk I want to make client upload statistics in batches before the merge

Should I take this one #197 ?

pbylicki commented:
@MaciejWiczk If you have time in the next couple of days then go ahead. Otherwise I can do this at the end of this week.

MaciejWiczk commented:
> @MaciejWiczk If you have time in the next couple of days then go ahead. Otherwise I can do this at the end of this week.

I will give it a go today or tomorrow, I will come back with either a PR or cry for help ;)

MaciejWiczk committed:
Modifying statistics_importer to send stats in batch per file (#200)

* Modifying statistics_importert to send stats in batch per file

* Fixed typing

* Making add_statistics into method for single list not list of lists

* Bumping fastapi to 0.46 since it requires pydantic above 1.0

* Making add_statistics into method for single list not list of lists (#201)

* Making add_statistics into method for single list not list of lists

* Bumping fastapi to 0.46 since it requires pydantic above 1.0

* Adding unit tests for add_statisticsand removing unused KeywordStatisticsList.one

* blacked

* Fixing locator

* bumping chromedriver version

* Changing p in into p[2]

* Trying sleep to let selenium load

* Freezing fastapi to 0.46-0.52

* Remove sleep

pbylicki committed:
Merge

pbylicki committed:
Rebuild UI
